### ISSUE TITLE
Only apply increment when isKeyFrame()

### ISSUE THREAD

**Issue by fabolhak** [artivis/laser_odometry#28]
Hello,
sorry for being so spammy, but I was wondering why the increment is applied to the current pose even though it may not be a keyframe. This results in a lot of noise in my laser odometry calculation (only checked CSM so far). When I execute the `posePlusIncrement()` only when `isKeyFrame()` is true, then I get a lot cleaner results.

So I propose a change in the process order:
```
process :
       - if first reading
           - initialize 
           - return
       - preProcessing
       - getIncrementPrior
       - processImpl  
       - is_keyframe = isKeyFrame(delta_pose) 
       - if is_keyframe is true
          - posePlusIncrement
          - isKeyFrame 
       - if is_keyframe is NOT true
          - isNotKeyFrame
       - postProcessing
```

My implementation looks like this:
https://github.com/fabolhak/laser_odometry/commit/b13a459a76ba2091d882fae74c552386cee9ac58

**Comment by artivis** [comment]
Hi,
thanks for your interest in this project. Let me start by saying that unfortunately I didn't have time to fully finish / wrap it up therefor all plugins may not be fully functional. I'm mostly using the [rf2o](https://github.com/artivis/laser_odometry_rf2o) plugin for my use case. The same manner I should improve the documentation as soon as possible.

The rational behind this execution scheme is actually inspired from the [csm node implementation](https://github.com/ccny-ros-pkg/scan_tools/blob/indigo/laser_scan_matcher/src/laser_scan_matcher.cpp). Let me summarize it here.  
- `processImpl` The scan matching is performed with the last scan selected as 'keyframe' and the last received scan. It thus estimates the displacement from the 'keyframe' pose (considered as identity) to the current one in the sensor frame. That is the `increment_` variable.
- `posePlusIncrement` First it transforms the estimated displacement `increment_` from the sensor frame to the base frame (`increment_in_base_`). It then compose the 'keyframe' pose with the new increment to obtain the current pose in the fixed frame (origin of the tracking)
 ( -> `fixed_to_base_ = fixed_to_base_kf_ * increment_in_base_` )
The pose is then expressed in the optional 'global frame' `fixed_origin_to_base_ = fixed_origin_ * fixed_to_base_` (by default `fixed_origin_ = I`).
- `updateKf` Finally, in case `increment_in_base_` is considered for creating a 'keyframe', the 'keyframe' pose is updated `fixed_to_base_kf_ = fixed_to_base_`.

Notice that the odometry message is constantly updated & published. Maybe should it be published only on new 'keyframe'. The noise you are observing may comes from this ?  
Please have a look at https://github.com/artivis/laser_odometry/commit/f133a3a85212b99a7348e08e33be7ee7bad39316.

While I cannot guarantee 100% that the implementation is error-proof, I have been using it successfully for a while now.

As for your modification fabolhak/laser_odometry@b13a459, just a remark. The variable `increment_` is in the sensor frame, making it sometime difficult to get your head around what should the 'keyframe' thresholds be. One the other hand, shall the laser frame be very different from the base frame, one should probably consider `increment_` rather than `increment_in_base_`.

I will 'play' with 'csm' as soon as possible see if I experience the same issue. In the mean time feel free to provide more feedback / PR :+1: .

**Comment by artivis** [comment]
This particular issue was addressed in #31 . Closing.
@fabolhak Please feel free to re-open it if needed.
